### PR TITLE
Api 42951 poa v2 zip code updates

### PR DESCRIPTION
**Pull request by rockwellwindsor-va**
## Summary
`zipCode` is only required now when the `countryCode` is 'US'
* Updates schema
* updates validations
	* Adjusts/add tests
* Updates documentation to reflect changes for zipCode

## Related issue(s)
[API-42951](https://jira.devops.va.gov/browse/API-42951)
## Testing done

- [x] *New code is covered by unit tests*
- [x] *Old tests adjusted*

#### Testing Notes
* Any 2122 & 2122a v2 submissions where the `countryCode` and `zipCode` are taken out/put in will test the behavior
Screenshot 2024-12-04 at 7 40 03 AM

## Screenshots
* 2122
Screenshot 2024-12-04 at 7 18 09 AM
Screenshot 2024-12-04 at 7 19 11 AM

* 2122a
Screenshot 2024-12-04 at 7 15 06 AM
Screenshot 2024-12-04 at 7 15 14 AM
Screenshot 2024-12-04 at 7 16 19 AM

## What areas of the site does it impact?
    modified:   modules/claims_api/app/controllers/claims_api/v2/veterans/power_of_attorney/base_controller.rb
	modified:   modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb
	modified:   modules/claims_api/config/schemas/v2/2122.json
	modified:   modules/claims_api/config/schemas/v2/2122a.json
	modified:   modules/claims_api/spec/requests/v2/veterans/power_of_attorney/2122_spec.rb
	modified:   modules/claims_api/spec/requests/v2/veterans/power_of_attorney/2122a_spec.rb
	modified:   modules/claims_api/spec/requests/v2/veterans/power_of_attorney/power_of_attorney_request_spec.rb
	modified:   modules/claims_api/app/swagger/claims_api/v2/dev/swagger.json
	modified:   modules/claims_api/app/swagger/claims_api/v2/production/swagger.json

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb [hunk 1]
```diff
@@ -40,6 +40,32 @@ def build_dependent_claimant_verification_service(veteran_participant_id:, user_
                                                             poa_code:)
       end
 
+      # the Claimant object is already being validatated below,
+      # so that code has been adjusted and it is being left out of this workflow
+      def validate_address_values
+        address_objects = { veteran: 'veteran', representative: 'representative' }
+
+        address_objects.each do |key, base|
+          address = form_attributes.dig(key.to_s, 'address')
```
Comment (tycol7) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb:49]:
```suggestion
        %w[veteran representative].each do |base|
          address = form_attributes.dig(base, 'address')
```

**Review by stiehlrod — APPROVED** [review]
tested V2 2122 & 2122a. LGTM

File: modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb [hunk 2]
```diff
@@ -40,6 +40,32 @@ def build_dependent_claimant_verification_service(veteran_participant_id:, user_
...
+
+        zip_code = address['zipCode']
+
+        if zip_code.blank?
+          collect_error_messages(
+            source: "/#{base}/address/zipCode",
+            detail: "If 'countryCode' is 'US' then 'zipCode' is required."
+          )
+        end
+      end
```
Comment (tycol7) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb:67]: Looks like we are duplicating logic:

https://github.com/department-of-veterans-affairs/vets-api/blob/751d05609fd423a144f0296f41e9b6c27000871a/modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb#L155-L162

Could we consolidate to one function that takes `address` and `base`, then refactor the claimant address validation to pass in `base: 'claimant'`?

File: modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb [hunk 3]
```diff
@@ -40,6 +40,32 @@ def build_dependent_claimant_verification_service(veteran_participant_id:, user_
                                                             poa_code:)
       end
 
+      # the Claimant object is already being validatated below,
+      # so that code has been adjusted and it is being left out of this workflow
+      def validate_address_values
```
Comment (tycol7) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb:45]: Could we rename the method to capture the comment? Something like:

```suggestion
      def validate_non_claimant_address_values
```

File: modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb [hunk 4]
```diff
@@ -78,7 +104,8 @@ def validate_address
           collect_error_messages(
             source: '/claimant/address/',
             detail: "If claimant is present 'address' must be filled in " \
-                    ' with required fields addressLine1, city, stateCode, country and zipCode'
+                    "with required fields addressLine1, city, stateCode and country. If the stateCode is 'US' " \
+                    'then zipCode is also required.'
```
Comment (tycol7) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb:108]:
```suggestion
                    "with required fields addressLine1, city, stateCode and countryCode. If the countryCode is 'US' " \
                    'then zipCode is also required.'
```

File: modules/claims_api/config/schemas/v2/2122a.json [hunk 5]
```diff
@@ -79,7 +78,7 @@
               "example": "USA"
             },
             "zipCode": {
-              "description": "Zipcode (First 5 digits) of the address.",
+              "description": "Zip code (First 5 digits) of the address.Required if 'country' is 'US'.",
```
Comment (tycol7) [inline modules/claims_api/config/schemas/v2/2122a.json:81]: Same comment as above (missing space between `address.Required`). Also, we refer to `country` and `countryCode` interchangeably throughout this PR. Should we stick to one?

**Review by tycol7 — COMMENTED** [review]
looking good so far. are we using `country` or `countryCode`? seems to be a mix of both throughout.

**Review by tycol7 — CHANGES_REQUESTED** [review]
we should resolve country vs countryCode before merging.

File: modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb (hunk 2 above)
Comment (rockwellwindsor-va) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb:67]: Ya good eye on that, updating now

**Comment by rockwellwindsor-va** [comment]
> looking good so far. are we using `country` or `countryCode`? seems to be a mix of both throughout.

Yep totally got some bleed over on that term from the other PR with that update in it, cleaning it up now

File: modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb [hunk 6]
```diff
@@ -40,6 +41,16 @@ def build_dependent_claimant_verification_service(veteran_participant_id:, user_
                                                             poa_code:)
       end
 
+      # the Claimant object is already being validatated below,
+      # so that code has been adjusted and it is being left out of this workflow
```
Comment (tycol7) [inline modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb:45]: We shouldnʼt need the comment now that the method name is self-descriptive.

**Review by tycol7 — DISMISSED** [review]
looks great!

**Review by tycol7 — APPROVED** [review]

File: modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb (hunk 6 above)
Comment (rockwellwindsor-va) [reply modules/claims_api/app/controllers/concerns/claims_api/v2/power_of_attorney_validation.rb]: ah, yep [cabc789](https://github.com/department-of-veterans-affairs/vets-api/pull/19665/commits/cabc78934041de80472ed1105d99ef02e8925aa1)

**Pull request merged**